File: morning_exercises/day14.py
from __future__ import annotations

class BSTNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
    
    @staticmethod
    def inorder(root: BSTNode | None, result: list, answer: list, k: int)->None:
        if root is None or result[0]==k: return

        BSTNode.inorder(root.left, result, answer, k)
        result[0]+=1
        if k==result[0]:
            answer[0]=root.val
            return
        BSTNode.inorder(root.right, result, answer, k)

    @staticmethod
    def ioListApproach(root: BSTNode | None, result: list)->None:
        if root is None: return

        BSTNode.ioListApproach(root.left, result)
        result.append(root.val)
        BSTNode.ioListApproach(root.right, result)

    @staticmethod
    def kthSmallest(root, k):
        result = [0]
        answer = [None]
        BSTNode.inorder(root, result, answer, k)
        return answer[0]
    
    @staticmethod
    def kthSmallestListApproach(root, k):
        result = []
        BSTNode.ioListApproach(root, result)
        return result[k-1]
    
root = BSTNode(4)
root.left = BSTNode(2)
root.left.left = BSTNode(1)
root.left.right = BSTNode(3)
print(BSTNode.kthSmallest(root,3))
print(BSTNode.kthSmallestListApproach(root,3))

File: morning_exercises/day14.cpp
#include <climits>
#include <memory>
#include <iostream>
#include <vector>

class BSTNode {
    public:
        int val = 0;
        std::shared_ptr<BSTNode> left = nullptr;
        std::shared_ptr<BSTNode> right = nullptr;
    BSTNode(int val): val(val) {}

    static void inOrderLA(std::shared_ptr<BSTNode> root, std::vector<int>& result, int k) {
        if (root==nullptr) return;

        inOrderLA(root->left, result, k);
        result.push_back(root->val);
        inOrderLA(root->right, result, k);
    }

    static int kthSmallestLA(std::shared_ptr<BSTNode> root, int k) {
        std::vector<int> result = {};
        inOrderLA(root, result, k);
        return result[k-1];
    }

    static void inOrder(std::shared_ptr<BSTNode> root, std::vector<int>& result, std::vector<int>& answer, int k) {
        if (root==nullptr || answer[0]!=INT_MIN) return;

        inOrder(root->left, result, answer, k);
        result[0]++;
        if (k==result[0]) {
            answer[0]=root->val;
            return;
        }
        inOrder(root->right, result, answer, k);
    }

    static int kthSmallest(std::shared_ptr<BSTNode> root, int k) {
        std::vector<int> result = {0}, answer = {INT_MIN};
        inOrder(root, result, answer, k);
        return answer[0];
    }
};

int main() {
    auto root = std::make_shared<BSTNode>(5);
    root->left = std::make_shared<BSTNode>(3);
    root->right = std::make_shared<BSTNode>(7);
    root->left->left = std::make_shared<BSTNode>(2);
    root->left->right = std::make_shared<BSTNode>(4);

    std::cout << BSTNode::kthSmallestLA(root, 3) << "\n";
    std::cout << BSTNode::kthSmallest(root, 3) << "\n";
}

